[클린코드 리액트 2기 신형기] 장바구니 미션 Step - 1

hyungkishin opened the pull request:
# Pull Request Template

## 1. 가이드

### 1-1. 배포
[장바구니](https://clean-shop-cart.netlify.app/)

### 1-2. 로컬 환경

```bash
# install
yarn

# run
yarn start

# test
yarn test
```

## 2. 리뷰 요청 ✍️
msw 테스트 작성한 방식이 맞는지 컨펌 부탁드려요

### 2-1. 어렵거나 아쉬웠던 점
1. msw 를 처음 작성해서 살짝 안드로메다 간것 같습니다.
2. 재사용 컴포넌트를 위해 styled-component 방식 을 상품목록 list 부터 그려나갔으나, css 모듈화 일찍 안하면 피본다는사실 늦게 알았습니다. ( 생각보다 찢기 힘들고 고난스러움 ) 


### 2-2. 나누고 싶은 고민

1. src/api 이부분은 실제 api 를 통신하는 부분 
    - src/api 디렉토리는 실제 API와 통신하는 코드가 포함된 디렉토리
2. src/mocks/api 이부분은 mock 용 api 를 따로 만들어 줘야 하는지 ?
    - src/mocks/api 디렉토리는 MSW를 사용하여 API를 가짜 응답으로 대체하는 목적으로 사용되는 mock API 코드가 포함된 디렉토리

결론적으로 실 api 파일과 mock api 파일 을 나눈것이 맞는 방향인지 고민 - 해결되었습니다 

### 2-3. 중점적으로 리뷰받고 싶은 부분
MSW, recoil, react-query 도입

### 2-4. 새로운 시도 혹은 미션을 통해 도전한 점
1. 감히 context-api 를 충분히 맛보았다 생각하여, 상태관리 라이브러리로 recoil과 react-query 를 도전해 보았습니다.
2. 라이브러리 쇼핑에 빠지면 안되는데... (그렇다고 무작정 도입한 부분은 아니였습니다 :) ) 
  - react-confirm-modal 도입

### 2-5. 도식화된 자료 혹은 작성된 이미지 형태의 설계 구조
이런 양식이 생겼다는걸 이제알아서 만들고 다시 PR 드릴까 하다가 리뷰후에 올리도록 하겠습니다...

## 3. 질문있어요 🙋
선택 요구사항은 리뷰를 받으면서 작성해 나가면 되는것인지 (다음 스텝과 연관있어보임)

## 4. 요구 사항 ✅
체크체크

### 4-1. 필수 요구사항
- [x] MSW 를 활용한 API mocking
  - [x] Endpoint만 변경하면 언제든 Real API를 바라볼 수 있다고 가정하고 상상합니다.
  - [x] Real API 없이 로컬에서만 동작하는 상태로 리뷰 받는 것이 기본 원칙입니다.

### SELF TEST  
- [x] npm run test (테스트가 통과하는지 )

### GNB
- [x] 로고를 누르면 상품목록 페이지로 이동한다.
- [x] 장바구니 버튼을 누르면 장바구니 페이지로 이동한다.
- [x] 주문목록 버튼을 누르면 주문목록 페이지로 이동한다.

### 상품목록
- [x] 상품들은 n x 4 레이아웃으로 보여진다.
- [x] 상품들에는 사진, 이름, 금액이 보여진다.
- [x] 장바구니 버튼을 클릭하면 (**)

### 4-2. 선택 요구사항 (심화)
- [x] 페이지에는 상품 사진, 이름, 금액 정보가 보여진다.
- [x] 장바구니 버튼을 클릭하면 장바구니 페이지로 이동한다.
- [ ] 장바구니 버튼을 클릭하면 해당 상품이 장바구니에 담긴다.

### 주문 상세
- [ ] 주문 정보가 보여진다.
  - [ ] 장바구니 버튼을 클릭하면, 해당 상품이 장바구니에 담기고 장바구니 이동 선택 모달이 보여진다.
    - [x] 모달 이 보여진다
    - [x] 장바구니 페이지로 이동한다
    - [ ] 해당 상품이 장바구니에 담긴다
  - [x] 장바구니 이동 버튼을 누르면 장바구니 페이지로 이동한다.

### UX/UI

- [ ] 반응형 레이아웃을 구현한다.
- [ ] 사용성 개선
  - [ ] 사용자를 위한 로딩 환경 개선
  - [ ] 상품이 없을 때와 같은 다양한 Edge Case 대응
  - [ ] 반응형 레이아웃 구현
  - [ ] 별도의 모바일 레이아웃 추가 제공
- [ ] 매출 증대 및 마케팅을 위해 별도의 기능 구현 (별도의 API 없음)
  - [ ] 브라우저 새로고침시 모든 상태 유지
  - [ ] 흐름을 고려한 맞춤 큐레이팅 상품 추천 기능
  - [ ] 구매 유도를 위한 상품 찜 페이지
- [ ] 매출 증대 및 마케팅을 위한 별도의 도구 추가
  - [ ] Google Analytics
  - [ ] Google Tag Manager

jinsoobyeon reviewed (changes requested):
안녕하세요. 태은님!
역시나 인상 깊은 코드를 보면서 많이 배웠습니다.

비동기 처리에 대한 테스트 및 예외 처리가 잘 이루어져 있고 다양한 엣지 케이스까지 고려하셔서 매우 안전한 상태관리로 보입니다!
리듀서 설계 또한 탄탄하게 하신 것으로 확인했습니다.
다양한 디바이스를 고려한 반응형 UI를 구현하셔서 유저 사용성에도 문제가 없어 보입니다.

다만 현재 인피니티 로딩으로 페이징을 구현하신 것으로 보이는데 스크롤이 되지 않아서 화면 밑의 상품이 보이지 않는 현상이 있습니다.
그리고 스타일드 컴포넌트를 사용하셨는데 공통 스타일 외에 컴포넌트마다 스타일 파일을 따로 작성하게 되면 불필요한 보일러 플레이트가 발생하지 않나 싶습니다. 예를 들어 컴포넌트에 반복적인 Styled가 들어가고 있습니다. 공통 스타일은 공통으로 빼서 Styled로 사용하고 컴포넌트에만 적용되는 스타일은 한 파일에 작성하는 것이 Styled에 대한 의미도 명확해지지 않나 생각합니다~

이상입니다~ 인피니티 스크롤 부분만 수정하시면 머지하겠습니다!

jinsoobyeon reviewed (approved):
안녕하세요 태은님!
태은님 말씀을 듣고 보니 제가 추상화되지도 않은 로직을 공용으로 분리하는 게 어떠냐는 의견을 드렸네요 ㅠㅠ 깊이 반성합니다.. 뭐라도 의견을 더 드려야 하겠다는 생각 때문에 이런 재앙?이 벌어진 것 같아요 ㅎㅎ 이번 계기를 바탕으로 더 나은 리뷰를 하도록 노력하겠습니다!

스크롤 부분에서는 마이너한 문제가 있는 게 스크롤을 맨 밑으로 내리고 큰 화면에서 작은 화면으로 화면 사이즈가 변경되면 변경된 화면에서는 화면 밑의 아이템이 렌더링 되지 않습니다 ㅠㅠ 시간이 되신다면 다음 step에서 수정해 보시는 것도 좋을 것 같아요~

이상 머지하겠습니다.

Commit:
feat: css 추가

Commit:
feat: cra 초기 setting

Commit:
feat: MSW lib 추가

Commit:
docs: 요구사항 체크

Commit:
feat: Header 컴포넌트 분리 및 with stitches/react

Commit:
feat: ItemList styling

Commit:
feat: nav styling

Commit:
feat: dbjson server 추가

Commit:
feat: axios 도입  productsPage Rendering

Commit:
feat: router 생성 ProductPage 연결

Commit:
feat: reset Css 추가

Commit:
feat: router 정리

Commit:
docs: 요구사항 체크

Commit:
feat: Nav Router 적용

Commit:
feat: icon 위치 변경, project default path 설정

Commit:
refactor: project import path 전체 수정

Commit:
refactor: 상품 목록 페이지 리펙토링 및 api 를 2번 call 하는 이유 확인후 store 생성준비

Commit:
refactor: 상품 목록 페이지 리펙토링

Commit:
feat: store > context, reducer 분리

Commit:
feat: server.js 를 바라보게 package.json 수정

Commit:
feat: css 민족 대이동 (도메인에 맞는 컴포넌트끼리)

Commit:
refactor: icon 경로 수정

Commit:
feat: carts 컴포넌트 추가 및 api 연동

Commit:
docs: 요구사항 체크

Commit:
feat: MSW 를 활용한 API mocking test

Commit:
docs: 요구사항 체크

Commit:
feat: route 정리

Commit:
docs: README update

Commit:
docs: 요구사항 체크

Commit:
refactor: mocks api > cart, product return data naming 변경

Commit:
refactor: 미사용 import 정리

Commit:
refactor: fileName 에 공백 제거

Commit:
refactor: actionType 네이밍 변경

woobottle commented on md/step-1.md:
```diff
@@ -0,0 +1,47 @@
+### 상품목록
+- [x] 상품들은 n x 4 레이아웃으로 보여진다.
+- [x] 상품들에는 사진, 이름, 금액이 보여진다.
+- [ ] 장바구니 버튼을 클릭하면 (**) <- comment ? 상세로 간다는 걸까요 
```
> 자유라는 의미입니다 ㅎㅎ
조금 더 친절하게 써볼까하다.. 상상력에 맡겨봤어요!

이런식으로 답변해주셨어요~~

woobottle commented on src/components/header/style.tsx:
```diff
@@ -0,0 +1,63 @@
+
+export const GlobalNavButton = styled("a", {
+  backgroundColor: '#ffffff',
+  border: '1px solid #222222',
```
색깔 변수들 중복되는 것들이 많아서 한곳에서 변수로 관리되면 어떻게 생각하실지 궁금해요!

woobottle commented on src/hooks/useRouter.tsx:
```diff
@@ -0,0 +1,22 @@
+    [navigate]
+  );
+
+  return {
```
👍🏻

woobottle commented on src/page/cart.html:
```diff
@@ -0,0 +1,176 @@
+<!DOCTYPE html>
+<html lang="en">
```
이 html파일들은 현재 쓰이지 않는 것일까요??

만약에 쓰이지 않는다면 제거 하는것은 어떻게 생각하실지 궁금해요!

woobottle commented on src/router/index.tsx:
```diff
@@ -0,0 +1,29 @@
+import { createBrowserRouter } from "react-router-dom";
+import CartPage from "page/cart";
+
+export const ROUTE = {
```
👍🏻

woobottle commented on src/store/context/ProductsContext.tsx:
```diff
@@ -0,0 +1,47 @@
+  const context = useContext(ProductsContext);
+
+  if (!context) {
+    throw new Error("useProductsContext must be used within a ProductsProvider");
```
👍🏻

woobottle commented on src/store/context/CartContext.tsx:
```diff
@@ -0,0 +1,47 @@
+
+  useEffect(() => {
+    fetchCarts();
+  }, [fetchCarts]);
```
이 코드의 의도가 궁금해요 
최초 로드시 받아오기 위함일까요?

hyungkishin replied:
페이지에서도 호출하고 있어 삭제하였습니다 . 
최초로 로드시 받아오기 위함이 목적이였으나 중복으로 제거 하였습니다

woobottle commented on src/page/list/index.tsx:
```diff
@@ -0,0 +1,39 @@
+
+  useEffect(() => {
+    handleFetchProducts();
+  }, [handleFetchProducts]);
```
이 부분으로 인해 api 호출 (페이지 레벨, context 구성쪽)이 두번씩 되어지고 있는것 같은데

의도 되어진것인지 궁금해요!

woobottle commented on src/page/cart/index.tsx:
```diff
@@ -0,0 +1,47 @@
+
+  useEffect(() => {
+    handleFetchProducts();
+  }, [handleFetchProducts]);
```
products 부분의 코멘트와 같은 내용이여요~

woobottle commented on src/__tests__/cart.test.js:
```diff
@@ -0,0 +1,32 @@
+beforeAll(() => server.listen());
+afterEach(() => server.resetHandlers());
+afterAll(() => server.close());
+
```
msw에서 단순히 응답을 주는것을 테스트 하기 보다는 

msw에서 받아온 데이터를 가지고 컴포넌트에서 잘 보여지는 지를 테스트 하는게 좋을것 같은데 어떻게 생각하실지 궁금해요!

프론트 개발자가 수행해야 하는 역할과 저희가 테스트로 안정성을 확보하고 싶은 부분은 api의 응답 결과가 아니라 그것을 어떻게 다루고 알맞게 보여주냐 이지 않을까 생각해요!

api에 대한 안전성 테스트도 중요할 수 있다고 생각해요  
하지만 개발/운영에서 테스트를 위해 실제 api요청 (db에 부하등)을 수행해야 하고 
누구의 역할인지 (백엔드/프론트엔드) 모호한 부분들이 있는것 같아요
이 모호한 부분들은 팀 바이 팀이지 않을까 생각이 들어요

hyungkishin replied:
크흐... 👍 
MSW 에서 받아온 데이터를 가지고 컴포넌트에서 잘 보여지는 지를 테스트 할것.

woobottle commented on src/mocks/api/cart.ts:
```diff
@@ -0,0 +1,5 @@
+export const getCarts = async () => {
+  const response = await fetch("/api/carts");
```
위와 같이 설정하신 이유를 여쭤보고 싶어요!!

제가 알고 있는 msw의 지식은 api의 주소가 바뀌는 것이 아니고 serviceWorker가 서버로 가야하는 요청을 훔쳐와 커스텀 데이터로 리턴해주는 것이 핵심이라고 생각해서요(프론트에서 백엔드로의 의존성 없이 병렬적으로 개발할 수 있게 해주는 요소)

hyungkishin replied:
동영상을 친절히 주신덕에 msw 를 이제야 이해한것 같습니다 .. 알려주신 msw 블로그 진짜 내용 너무좋네요 감사합니다.

woobottle reviewed (changes requested):
제일 먼저 수행해주셔서 고생 많으셨어요!! 👍🏻
좀 더 리뷰를 잘 수행하기 위해서 몇 가지를 더 여쭤보고 싶어요
답변 해주시는 대로 좀 더 저랑 티키타카 해보시죠!!

--------

1. msw세팅은 해주셨지만 실제 개발시에는 사용하지 않은것으로 확인되는데 맞을까요??
2. contextApi를 선택하신 이유가 궁금해요 (카드 미션쪽에서 뚫어서 사용하시는 것인지, 기술적 선택을 하신것인지 궁금해요(여러 페이지 간에 공유 되어야 한다 등등 형기님의 의도가 궁금해요)) 
3. contextapi에서 어느 부분을 중점적으로 이야기 나누고 싶으신지 궁금해요!
4. api 핸들러 부분은 어느 부분을 뜻하는 것일지 알 수 있을까요? (handleFetchProducts, handleFetchCart 이 부분일까요??)
5. 나누고 싶은 고민의 1번은 어떤 고민일지 여쭤 보고 싶어요
> 2-2 나누고 싶은 고민
 src/api 이부분은 실제 api 를 통신하는 부분
src/api 디렉토리는 실제 API와 통신하는 코드가 포함된 디렉토리


----------

> 2-2. 나누고 싶은 고민
src/mocks/api 이부분은 mock 용 api 를 따로 만들어 줘야 하는지 ?

위의 부분은 코멘트에 제 생각 남겨두었어요!

---------

첨부한 영상은 제가 msw를 사용했던 방식을 형기님의 프로젝트에 도입한 영상 인데요 
혹시 도움이 될까 첨부합니다

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/50283326/225077239-e039361e-75cd-467f-8f1e-93ac7323baf5.gif)

------------

[msw 참고용 블로그1](https://blog.mathpresso.com/msw%EB%A1%9C-api-%EB%AA%A8%ED%82%B9%ED%95%98%EA%B8%B0-2d8a803c3d5c)
[msw 참고용 블로그2](https://www.daleseo.com/mock-service-worker/)

Commit:
refactor: msw mocking, App.tsx 에 적용 및 코드리뷰 반영

Commit:
delete: 불필요 env 파일 제거

Commit:
refactor: global Header 수정

장바구니, 주문목록 선택시 페이지 랜딩

Commit:
feat: react-query, recoil 도입

사유: context를 현재 로그인한 유저, 테마, 선호하는 언어 등의 데이터 공유 시에 사용할 것을 추천
본 프로젝트에선 부적합하다고 판단.
물론 쓸라면 쓸순있겠지만 왜 사람들이 그렇게 recoil 을 찬양할까 비교글을 보아하니 단순하고 + 데이터를 atom 단위로 보는것과 여러가지 장점이 존재한다.
반면 현재 context-api 는 Provider와 컴포넌트 트리 노드 사이에 강한 커플링이 생겨버려서 쓸만큼 써봤고 지겨웠고. recoil 로 탈출하려고 한다.
react-query 는 서버 상태 가져오기, 캐싱, 동기화 및 업데이트를 쉽게 만들어주기 때문에 loading spinner 과 error 상태를 좀더 쉽게 뽑아먹을수 있다는 장점과
제일 중요한 캐싱 ... 동기화 이부분을 사용해보려고 한다.

Commit:
feat: react-query, recoil 적용

Commit:
chore: 파일 naming 변경 및, product data fetching querycache 1분 설정

Commit:
feat: react-query devtools 추가

Commit:
refactor: pageContent useEffect 내부 로직으로 이동

Commit:
refactor: json server start 구동 명령어로 merge

Commit:
feat: orderTest 및 hooks 작성

Commit:
feat: header -> global Header 이름 변경

Commit:
refactor: cart.ts 타이핑

Commit:
feat: orderList Page, DetailPage 작성

Commit:
refactor: common/ui Header, icon 공통화

Commit:
feat: modal component 추가 및 공통화

Commit:
refactor: common/ui, modal 분리

Commit:
feat: step-1 요구사항 체크 및, 상품상세 장바구니 선택시 이동

Commit:
refactor: 미사용 html 이동

woobottle commented on src/page/product/list/item.tsx:
```diff
@@ -0,0 +1,50 @@
+
+  return (
+    <div onClick={(e) => handleMoveToDetail(e, item.id)}>
+      <Image src={`${item.imageUrl}`} alt="PET 보틀-정사각(420ml)" />
```
놓칠 수 있는 부분(alt="")이라 생각하는데 디테일 하시군요!! 👍🏻

hyungkishin replied:
이부분 수정해야는 부분인데 :) 감사합니다

woobottle commented on src/page/product/list/index.tsx:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    if (data) {
+      setProducts(data);
```
전체 상품을 불러올때마다 recoil로 관리되어 지는것 으로 파악하였는데요
혹시 어떤 것을 염두에 두신것일지 여쭤볼 수 있을까요??

hyungkishin replied:
어제 강의에서 관리해야할 상태가 늘어나는것이 좋지만은 않다라고 말씀하신 부분이 콕찔립니다...
해당 컴포넌트에서 useState 로 관리되어도 될것같아 이부분은 수정하겠습니다 :)

woobottle commented on src/hooks/product/index.ts:
```diff
@@ -0,0 +1,31 @@
+
+export function useSelectedProduct(productId: number): ProductItem | undefined {
+  const productList = useRecoilValue(productListState);
+  const selectedProductLoadable = useRecoilValueLoadable(productListState);
```
useRecoilValueLoadable 훅을 사용하신 것은 어떤 것을 의도에 두셨을지 궁금해요!!

hyungkishin replied:
아!!! 이부분 너무나도 피드백 받고싶었던 부분이였습니다.

productListState 의 상태의 값을 가지기 전까지는 useRecoilValue 와 useRecoilValueLoadable 는 거의 비슷하다고 볼수있으나,
대기시간을 처리하는 부분에서 차이가 있는걸로 알고있슴돠.

useRecoilValue 는 값을 사용할 수 있을 때까지 렌더링을 차단하기때문에, UI  랜더링이 지연될 수 있는걸로 알고있습니다.
useRecoilValueLoadable 는 현재 상태를 확인하는 동안 다른 UI 를 랜딩 할 수 있는 점이 달라 사용하였습니다.

결국 컴포넌트는 productListState가 값을 가지기 전에도 즉시 UI를 렌더링할 수 있게되고, 동시에 productListState의 상태를 확인하여 로딩 인디케이터나 로딩 바를 표현할 수 있다는 점으로 되겠습니다. 

사용자 경험을 높히는 의도가 있었습니다.

제가 잘못알고 있는 부분이 있거나 미흡한 부분이 있다면 지적 부탁드립니다 ㅠㅠ

woobottle replied:
사용자 경험을 높이시려는 의도는 너무 좋은것 같고 저도 배워야할 것들인것 같아요!!

제가 알고 있는 useRecoilValueLoadable도 큰 틀에서 형기님의 지식과 다르지 않아요!!
하지만 그 쓰임이 현재 적절한지에 대해서 이야기 해보고 싶어요

loading or hasValue를 사용해야 하는 부분들은 비동기 로직과 연관이 있을것 같아요
하지만 현재의 쓰임에서 비동기 로직등을 확인할 수 없어 여쭤보고 싶었어요!!

예를들어 
useRecoilValueLoadable(비동기 Selector)를 사용하고
컴포넌트에서는 이 데이터 로딩 상태를 보고 로딩, 성공, 실패 등을 표시해 줄 수 있지 않을까 했었어요~~

여담으로 
저는 코딩에서 누가 누구를 지적하고 이러한 것들은 할 수 도 없고 하는 것도 좋지 않다고 생각해요!!
같이 발전해 나가는 것이라 생각하고 저희가 시너지를 낼 수 있었으면 좋겠어요~~

woobottle commented on src/hooks/product/index.ts:
```diff
@@ -0,0 +1,31 @@
+  // https://velog.io/@aeong98/React.js-React-Query
+  const { data, isLoading, isError } = useQuery<ProductItem[]>(PRODUCTS_QUERY_KEY, getProducts, {
+    staleTime: 60 * 1000, // 1분
+    keepPreviousData: true, // 이전 데이터를 보여줌
```
이 옵션을 사용하신 것은 ux를 고려하시려고 추가하신 것일지 궁금해요~!!

hyungkishin replied:
막상 구현해 보고나니 조금 이상하네요 ㅋㅋㅋ...

서버쪽에 캐시가 되어있겠지만 :)  read 비용을 최소화 하고자 하여 추가해둔 부분이였습니다

woobottle reviewed (changes requested):
안녕하세요 형기님 
이번 미션은 머지해도 충분하다 생각되어요!!

1. 코멘트 몇개를 남겨두었어요~!!
2. 더 나은 리뷰를 위해 지난번 리뷰때 제가 남긴 질문들에 대한 답변을 요청드리고 싶어요! [링크](https://github.com/next-step/react-shopping-cart/pull/1#pullrequestreview-1339599876)

1,2 번에 대한 사항들만 확인되시면 바로 머지 할게요~~

hyungkishin commented:
💯 꼼꼼한 리뷰 💯  너무나 감사드립니다 🥹 👍

woobottle reviewed (approved):
코멘트 해주신 것에 리코멘트를 달아 두었어요~~

이전에 공유 드린것과 같이 이번 미션은 머지할게요~
바쁘신 일정 중에 빠르게 수행해주셔서 감사합니다!!
